libmexclass/cpp/CMakeLists.txt

```diff
@@ -48,111 +86,136 @@ target_link_libraries(${LIBMEXCLASS_LIBRARY_NAME} PRIVATE ${Matlab_MEX_LIBRARY})
+
+    find_package(Matlab REQUIRED)
+
+    matlab_add_mex(NAME ${LIBMEXCLASS_CLIENT_MEX_GATEWAY_NAME}
+                   SRC ${LIBMEXCLASS_CLIENT_MEX_GATEWAY_SOURCES}
+                   OUTPUT_NAME "gateway"
+                   LINK_TO mexclass
+                           ${LIBMEXCLASS_CLIENT_MEX_GATEWAY_CLIENT_PROXY_LIBRARY})
+    # Set compiler features to require C++17.
+    target_compile_features(${LIBMEXCLASS_CLIENT_MEX_GATEWAY_NAME} PRIVATE cxx_std_17)
+    # Include directories (i.e. header files).
+    get_target_property(LIBMEXCLASS_INCLUDE_DIRS mexclass INCLUDE_DIRECTORIES)
+    target_include_directories(${LIBMEXCLASS_CLIENT_MEX_GATEWAY_NAME} PRIVATE ${LIBMEXCLASS_INCLUDE_DIRS})
+    # TODO: Decide whether to automatically add client library include directories.
+    get_target_property(LIBMEXCLASS_CLIENT_PROXY_LIBRARY_INCLUDE_DIRS ${LIBMEXCLASS_CLIENT_MEX_GATEWAY_CLIENT_PROXY_LIBRARY} INCLUDE_DIRECTORIES)
+    target_include_directories(${LIBMEXCLASS_CLIENT_MEX_GATEWAY_NAME} PRIVATE ${LIBMEXCLASS_CLIENT_PROXY_LIBRARY_INCLUDE_DIRS})
+
+    target_set_rpath_relocatable(NAME ${LIBMEXCLASS_CLIENT_MEX_GATEWAY_NAME})
+endmacro()
+
+# ############
 # Installation
-# ########################
-
-# Copy ${LIBMEXCLASS_MATLAB_SOURCES} first so that ${CMAKE_CURRENT_BINARY_DIR}/+libmexclass/+proxy/ folder is there to copy binaries to.
-install(CODE "MESSAGE(\"Copying LIBMEXCLASS_MATLAB_SOURCES to packaged folder CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_DIR...\")")
-install(DIRECTORY ${LIBMEXCLASS_MATLAB_SOURCES} DESTINATION ${CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_ROOT_DIR})
-
-# Copy the binaries to ${CMAKE_CURRENT_BINARY_DIR}/+libmexclass/+proxy/ packaged folder.
-install(CODE "MESSAGE(\"Copying libmexclass binaries from ACTUAL_LIBMEXCLASS_BINARIES_DIR to packaged folder CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_DIR...\")")
-install(FILES ${INSTALL_BINARIES_LIST} DESTINATION ${CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_DIR})
-
-# Copy CUSTOM_PROXY_RUNTIME_LIBRARIES to ${CMAKE_CURRENT_BINARY_DIR}/+libmexclass/+proxy/ packaged folder.
-if(DEFINED CUSTOM_PROXY_RUNTIME_LIBRARIES)
-install(CODE "MESSAGE(\"Copying CUSTOM_PROXY_RUNTIME_LIBRARIES to packaged folder CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_DIR...\")")
-install(FILES ${CUSTOM_PROXY_RUNTIME_LIBRARIES} DESTINATION ${CMAKE_PACKAGED_LIBMEXCLASS_INSTALL_DIR})
-endif()
+# ############
+
+macro(libmexclass_client_install)
+    set(options)
+    set(oneValueArgs NAME CLIENT_PROXY_LIBRARY_NAME CLIENT_MEX_GATEWAY_NAME DESTINATION)
+    set(multiValueArgs SOURCES)
+    cmake_parse_arguments(LIBMEXCLASS_CLIENT_INSTALL "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    install(TARGETS mexclass
+                    ${LIBMEXCLASS_CLIENT_INSTALL_CLIENT_PROXY_LIBRARY_NAME}
+                    ${LIBMEXCLASS_CLIENT_INSTALL_CLIENT_MEX_GATEWAY_NAME}
+            DESTINATION ${LIBMEXCLASS_CLIENT_INSTALL_DESTINATION}/+libmexclass/+proxy)
```

Comment:
Should we set `+libmexclass/+proxy` to a variable so that in this macro it's hardcoded only once?

Reply:
Good idea! Fixed.